discovery_client.go

```diff
@@ -0,0 +1,148 @@
+package rendezvous
+
+import (
+	"context"
+	"github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+type rendezvousDiscoveryClient struct {
+	rp        RendezvousPoint
+	peerCache sync.Map //is a map[string]discoveredPeerCache
+	rng       *rand.Rand
+	rngMux    sync.Mutex
+}
+
+type discoveredPeerCache struct {
+	cachedRegs map[peer.ID]*Registration
+	cookie     []byte
+	mux        sync.Mutex
+}
+
+func NewRendezvousDiscoveryClient(host host.Host, rendezvousPeer peer.ID) discovery.Discovery {
+	rp := NewRendezvousPoint(host, rendezvousPeer)
+	return &rendezvousDiscoveryClient{rp, sync.Map{}, rand.New(rand.NewSource(rand.Int63())), sync.Mutex{}}
+}
+
+func (c *rendezvousDiscoveryClient) Advertise(ctx context.Context, ns string, opts ...discovery.Option) (time.Duration, error) {
+	// Get options
+	var options discovery.Options
+	err := options.Apply(opts...)
+	if err != nil {
+		return 0, err
+	}
+
+	ttl := options.Ttl
+	var ttlSeconds int
+
+	if ttl == 0 {
+		ttlSeconds = 7200
+	} else {
+		ttlSeconds = int(math.Round(ttl.Seconds()))
+	}
+
+	if returnedTTL, err := c.rp.Register(ctx, ns, ttlSeconds); err != nil {
+		return 0, err
+	} else {
+		return returnedTTL, nil
+	}
+}
+
+func (c *rendezvousDiscoveryClient) FindPeers(ctx context.Context, ns string, opts ...discovery.Option) (<-chan peer.AddrInfo, error) {
```

Comment:
you are taking and releasing the lock all the time here, better to just do it once and have the whole method under lock.

Reply:
I guess that's fine. It's super unfortunate that we have an interface that returns channel that is synchronous (sort of an implicit contract break).

Locking the cache during `Discover` may end up increasing latency to return. However, it saves us from a few annoying corner cases, so fine by me.